docs/okeps/okep-4380-network-qos.md

```diff
@@ -0,0 +1,404 @@
+# OKEP-4380: Network QoS Support
+
+* Issue: [#4380](https://github.com/ovn-org/ovn-kubernetes/issues/4380)
+
+## Problem Statement
+
+The workloads running in Kubernetes using OVN-Kubernetes as a networking backed might have different requirements in handling network traffic. For example video streaming application needs low latency and jitter whereas storage application can tolerate with packet loss. Hence NetworkQoS is essential in meeting these SLAs to provide better service quality.
+
+The workload taffic can be either east west (pod to pod traffic) or north south traffic (pod to external traffic) types in a Kubernetes cluster which is limited by finite bandwidth. So NetworkQoS must ensure high priority applications get the necessary NetworkQoS marking so that it can prevent network conjestion.
+
+## Goals
+
+- Provide a mechanism for users to set DSCP marking on east west and north south traffic.
+- Provide a mechanism for users to set Metering on east west and north south traffic.
+- Make NetworkQoS CRD fields to be more extensible.
+
+## Non-Goals
+
+- Ingress Network QoS.
+
+- Consolidating with current `kubernetes.io/egress-bandwidth` and `kubernetes.io/ingress-bandwidth` annotations.
+Nonetheless, the work done here does not interfere with the current bandwidth NetworkQoS mechanism.
+
+- The DSCP marking does not need to be handled or acted upon by OVN-Kubernetes, just added to selected headers.
+
+## Introduction
+
+To enable NetworkQoS, we will use Differentiated Services Code Point (DSCP) which allows us to classify packets by setting a 6-bit field in the IP header, effectively marking the priority of a given packet relative to other packets as "Critical", "High Priority", "Best Effort" and so on.
+
+## User-Stories/Use-Cases
+#### Story 1
+
+As a user of OVN-Kubernetes, I should be able to mark egress traffic coming from a specific namespace with a valid DSCP value.
+
+#### Story 2
+
+As a user of OVN-Kubernetes, I should be able to apply bandwidth limit (rate and burst) for a pod traffic originating from a specific namespace.
+
+#### Story 3
+
+As a user of OVN-Kubernetes, I should be able to create multiple NetworkQoS objects per namesapce containing rules with different priorities.
+
+#### Story 4
+
+The NetworkQoS Status field must reflect observed state of the object covering entire Kubernetes cluster.
+
+## Proposed Solution
+
+By introducing a new CRD `NetworkQoS`, users could specify a DSCP value for packets originating from pods on a given namespace heading to a specified CIDR, Protocol and Port. This
```

Comment:
In addition to CIDR, we should also say that we can set QoS to Pods as well using NSSelector or PodSelector

Reply:
done.